### PR TITLE
[Week 1] Add “Breaking API Changes“ example

### PR DESCRIPTION
**Pull request by mxdvl**
## What does this change?

Add a small CAPI search rendering project, parsing JSON responses, demonstrating the benefits of strict types.

## How to test

`npm run week-1-breaking-api-changes`

## How can we measure success?

The students will tell…

## Have we considered potential risks?

This may introduce search at the Guardian…

## Images

image

## Accessibility

-   [ ] [Tested with screen reader](https://github.com/guardian/accessibility/blob/main/people-and-technology/03-visual.md#screen-reader)
-   [X] [Navigable with keyboard](https://github.com/guardian/accessibility/blob/main/people-and-technology/02-physical.md#keyboard)
-   [X] [Colour contrast passed](https://github.com/guardian/accessibility/blob/main/people-and-technology/03-visual.md#contrast)
-   [X] [The change doesn't use only colour to convey meaning](https://github.com/guardian/accessibility/blob/main/people-and-technology/03-visual.md#use-of-colour)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  ],
+  "author": "",
+  "license": "MIT",
+  "bugs": {
+    "url": "https://github.com/guardian/typescript-school/issues"
+  },
+  "homepage": "https://github.com/guardian/typescript-school#readme",
+  "devDependencies": {
+    "typescript": "5.1.6",
+    "zod": "3.21.4"
```
Comment (JamieB-gu) [inline package.json:29]: What will this be used for?

Comment (mxdvl) [reply package.json:29]: Parsing CAPI JSON responses as part of week 1. Any other suggestion for an alternative?

**Review by JamieB-gu — DISMISSED** [review]

**Review by dskamiotis — APPROVED** [review]
This looks sensible and useful to me 👍

File: package.json (hunk 1 above)
Comment (mxdvl) [reply package.json]: See `search` method: https://github.com/guardian/typescript-school/pull/1/files#diff-2290a769a57bc280431b38ab6bd77b459e3c177dbb9c67c6c2ce047b1f87e905R26

File: 1-breaking-API-changes/server.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
+import { serve } from '@hono/node-server';
+import { Hono } from 'hono';
+import { search } from './capi.ts';
+
+const app = new Hono();
```
Comment (JamieB-gu) [inline 1-breaking-API-changes/server.ts:5]: In the interests of limiting the surface area of new concepts people have to learn, what do you think about using Node's built-in HTTP server[^1] for this?

[^1]: https://nodejs.org/api/http.html#httpcreateserveroptions-requestlistener

File: package.json [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+	},
+	"repository": {
+		"type": "git",
+		"url": "git+https://github.com/guardian/typescript-school.git"
+	},
+	"keywords": [
+		"typescript"
+	],
+	"author": "",
+	"license": "MIT",
```
Comment (JamieB-gu) [inline package.json:25]: Do we normally use Apache 2.0? 

https://github.com/guardian/dotcom-rendering/blob/main/LICENSE

https://github.com/guardian/dotcom-rendering/blob/ae6c83c7d36f1b661c02fdde23bcc08b2a4b52ed/package.json#L5

File: package.json [hunk 4]
```diff
@@ -0,0 +1,44 @@
+{
+	"name": "typescript-school",
+	"description": "🧑‍🎓🧑‍🏫🧑‍🎓 Resources for learning more about a common language at the Guardian: TypeScript ",
+	"type": "module",
+	"private": true,
+	"scripts": {
+		"lint": "npm run lint:package && npm run lint:prettier",
+		"lint:prettier": "prettier . --write",
+		"lint:package": "node ./node_modules/@guardian/package-linter/esm/cli.js ./package.json",
```
Comment (JamieB-gu) [inline package.json:9]: What does this do?

File: tsconfig.json [hunk 5]
```diff
@@ -0,0 +1,9 @@
+{
+	"extends": "@guardian/tsconfig",
+	"compilerOptions": {
+		"checkJs": true,
+		"allowImportingTsExtensions": true,
+		"moduleResolution": "Bundler"
+	},
+	"exclude": ["node_modules"]
+}
```
Comment (JamieB-gu) [inline tsconfig.json:9]: Is it worth adding comments explaining what each of these do, so people browsing the code can learn how it works?

**Review by JamieB-gu — COMMENTED** [review]
🎓 

Regarding the questions about mapping/using specific types for some of the fields in `capi.ts`: I'm not sure whether it's a good idea or not, I can see it having one of two consequences:

1. It's useful to see how TS allows you to be very precise with the models you're working with, describing the problem domain precisely.
2. It's confusing to include this extra complexity.

What do you think?

File: 1-breaking-API-changes/capi.ts [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+const fields = zod.object({
+	thumbnail: zod.string().optional(),
+	trailText: zod.string().optional(),
+	byline: zod.string().optional(),
+});
+
+type Result = zod.output<typeof result>;
+const result = zod.object({
+	pillarId: zod.string().optional(),
+	webPublicationDate: zod.string(),
```
[the remark this answers is not in the document]
Comment (mxdvl) [reply 1-breaking-API-changes/capi.ts:13]: Indeed, I did so in #3

File: 1-breaking-API-changes/capi.ts [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+const fields = zod.object({
+	thumbnail: zod.string().optional(),
+	trailText: zod.string().optional(),
+	byline: zod.string().optional(),
+});
+
+type Result = zod.output<typeof result>;
+const result = zod.object({
+	pillarId: zod.string().optional(),
```
[the remark this answers is not in the document]
Comment (mxdvl) [reply 1-breaking-API-changes/capi.ts:12]: Indeed, which I've done in #3

File: 1-breaking-API-changes/capi.ts [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+export const search = async (q: string | null): Promise<Result[]> => {
+	if (!q) return [];
+
+	const params = new URLSearchParams({
+		q,
+		orderBy: 'newest', // not "relevance"
+		'page-size': String(24),
+		'show-fields': ['thumbnail', 'trailText', 'byline'].join(','),
+		'api-key': 'test',
```
[the remark this answers is not in the document]
Comment (mxdvl) [reply 1-breaking-API-changes/capi.ts:34]: It's [used in the open-platform documentation](https://open-platform.theguardian.com/documentation/search)

File: package.json (hunk 3 above)
Comment (mxdvl) [reply package.json:25]: I never know for sure…

File: tsconfig.json (hunk 5 above)
Comment (mxdvl) [reply tsconfig.json:9]: There’s intellisense already for `tsconfig, did you have anything else in mind?

File: package.json (hunk 3 above)
Comment (mxdvl) [reply package.json]: But I’ve picked it now :)

**Review by JamieB-gu — APPROVED** [review]

File: 1-breaking-API-changes/server.ts (hunk 2 above)
Comment (mxdvl) [reply 1-breaking-API-changes/server.ts:5]: Fixed in #14

**Pull request merged**